internal/ethapi/celo_block.go

```diff
@@ -0,0 +1,166 @@
+}
+
+// PopulatePreGingerbreadHeaderFields populates the baseFee and gasLimit fields of the header for pre-gingerbread blocks
+func PopulatePreGingerbreadHeaderFields(ctx context.Context, backend CeloBackend, header *types.Header) *types.Header {
+	isGingerbread := backend.ChainConfig().IsGingerbread(header.Number)
+	if isGingerbread {
+		return header
+	}
+
+	var (
+		baseFee  *big.Int
+		gasLimit *big.Int
+	)
+
+	fields, err := rawdb.ReadPreGingerbreadAdditionalFields(backend.ChainDb(), header.Hash())
+	if fields != nil {
+		if fields.BaseFee != nil && fields.BaseFee.BitLen() != 0 {
+			baseFee = fields.BaseFee
+		}
+		if fields.GasLimit != nil && fields.GasLimit.BitLen() != 0 {
+			gasLimit = fields.GasLimit
+		}
+	} else {
+		if err != nil {
+			log.Debug("failed to read pre-gingerbread fields", "err", err)
+		}
+
+		// If the record is not found, get the values and store them
+		baseFee, err = retrievePreGingerbreadBlockBaseFee(ctx, backend, header.Number)
+		if err != nil {
+			log.Debug("Not adding baseFeePerGas to RPC response, failed to retrieve gas price minimum", "block", header.Number.Uint64(), "err", err)
+		}
+
+		gasLimit = retrievePreGingerbreadGasLimit(backend, header.Number)
+
+		if baseFee != nil || gasLimit != nil {
+			err = rawdb.WritePreGingerbreadAdditionalFields(backend.ChainDb(), header.Hash(), &rawdb.PreGingerbreadFields{
+				BaseFee:  baseFee,
+				GasLimit: gasLimit,
+			})
+			if err != nil {
+				log.Debug("failed to write pre-gingerbread fields", "err", err)
+			}
+		}
+	}
+
+	if baseFee != nil {
+		header.BaseFee = baseFee
+	}
+	if gasLimit != nil {
+		header.GasLimit = gasLimit.Uint64()
+	}
+
+	return header
+}
+
+// retrievePreGingerbreadGasLimit retrieves a gas limit at given height from hardcoded values
+func retrievePreGingerbreadGasLimit(backend CeloBackend, height *big.Int) *big.Int {
+	if backend.ChainConfig().ChainID == nil {
+		log.Debug("Not adding gasLimit to RPC response, unknown network")
+		return nil
+	}
+
+	limits, ok := params.PreGingerbreadNetworkGasLimits[backend.ChainConfig().ChainID.Uint64()]
+	if !ok {
+		log.Debug("Not adding gasLimit to RPC response, unknown network", "chainID", backend.ChainConfig().ChainID)
+		return nil
+	}
+
+	return new(big.Int).SetUint64(limits.Limit(height))
+}
+
+// retrievePreGingerbreadBlockBaseFee retrieves a base fee at given height from the previous block
+func retrievePreGingerbreadBlockBaseFee(ctx context.Context, backend CeloBackend, height *big.Int) (*big.Int, error) {
+	if height.Cmp(common.Big0) <= 0 {
+		return nil, nil
+	}
+
+	prevHeight := height.Uint64() - 1
+
+	prevBlock, err := backend.BlockByNumber(ctx, rpc.BlockNumber(prevHeight))
+	if err != nil {
+		return nil, err
+	}
+	if prevBlock == nil {
+		return nil, fmt.Errorf("block #%d not found", height.Int64())
+	}
+
+	prevReceipts, err := backend.GetReceipts(ctx, prevBlock.Hash())
+	if err != nil {
+		return nil, err
+	}
+
+	numTxs, numReceipts := len(prevBlock.Transactions()), len(prevReceipts)
```

Comment:
I wondering about this, why should the event just be included in a system transaction. As fas as I understand any update is going through governance and when approved is processed as a normal transaction.

Also, wouldn't the `baseFeePerGas` just be added to epoch blocks with the check below?

Reply:
Our base fee per gas was dynamically calculated by the gas price minimum contract at the end of each block, as such each block before gingerbread (except for genesis) contained a receipt, with the value calculated.